webhook/webhook.go

```diff
@@ -206,16 +221,38 @@ func (wh *Webhook) Run(stop <-chan struct{}) error {
 
 	select {
 	case <-stop:
-		if err := server.Close(); err != nil {
+		// Start failing readiness probes immediately.
+		logger.Debug("Starting to fail readiness probes...")
+		close(wh.stopCh)
+
+		// Wait for a grace period for the above to take effect and this Pod's
+		// endpoint to be removed from the webhook service's Endpoints.
+		// For this to be effective, it must be greater than the probe's
+		// periodSeconds times failureThreshold by a margin suitable to
+		// propagate the new Endpoints data across the cluster.
+		time.Sleep(wh.gracePeriod)
+
+		if err := server.Shutdown(context.Background()); err != nil {
```

Comment:
I think this is (was) incorrect. We probably need to `eg.Wait()` in any case.

Reply:
We `eg.Wait()` below (hasn't changed), which seems right because the go routine we launched won't return before Shutdown (or Close) is called.

Reply:
Clarified on chat that this is about the error path.  I tweaked the code so that the shutdown is done via `eg.Go` so that the trailing `eg.Wait` happens on all paths and we wait for all outstanding go routines to terminate.